libsplinter/src/peer/notification.rs

```diff
@@ -61,6 +63,51 @@ impl Iterator for PeerNotificationIter {
     }
 }
 
+pub type SubscriberId = usize;
+pub(super) type Subscriber =
+    Box<dyn Fn(PeerManagerNotification) -> Result<(), Box<dyn std::error::Error>> + Send>;
+
+/// Responsible for broadcasting peer manager notifications.
+pub(super) struct SubscriberMap {
+    subscribers: HashMap<SubscriberId, Subscriber>,
+    next_id: SubscriberId,
+}
+
+impl SubscriberMap {
+    pub fn new() -> Self {
+        Self {
+            subscribers: HashMap::new(),
+            next_id: 0,
+        }
+    }
+
+    pub fn broadcast(&mut self, notification: PeerManagerNotification) {
+        let mut failures = vec![];
+        for (id, callback) in self.subscribers.iter() {
+            if let Err(err) = (*callback)(notification.clone()) {
+                failures.push(*id);
+                debug!("Dropping subscriber ({}): {}", id, err);
+            }
+        }
+
+        for id in failures {
+            self.subscribers.remove(&id);
```

Comment:
Why can't this be done in the `if let Err...`?

Reply:
An immutable borrow occurs above